lib/goal.ex

```diff
@@ -793,6 +824,65 @@ defmodule Goal do
     end)
   end
 
+  defp validate_array_basic_field(changes, field, schema, type, changeset) do
+    params = Map.get(changes, field)
+    rules = Map.get(schema, field)
+    item_rules = Keyword.get(rules, :rules)
+
+    if item_rules && is_list(params) do
+      case reduce_and_validate_array_basic_fields(item_rules, type, field, params) do
+        {:valid, changes} ->
+          changeset
+          |> put_in([Access.key(:changes), Access.key(field)], Enum.reverse(changes))
+          |> Map.update!(:valid?, &Kernel.&&(&1, true))
+
+        {:invalid, errors} ->
+          changeset
+          |> Map.update!(:errors, &(errors ++ &1))
+          |> Map.put(:valid?, false)
+      end
+    else
+      changeset
+    end
+  end
+
+  @placeholder :inner
+  defp reduce_and_validate_array_basic_fields(schema, type, field, params) do
+    schema = %{@placeholder => [{:type, type} | schema]}
+
+    Enum.reduce(params, {:valid, []}, fn params, {status, acc} ->
+      changeset = build_changeset(schema, %{@placeholder => params})
+
+      case {status, changeset} do
+        {:valid, %Changeset{valid?: true, changes: %{@placeholder => inner_changes}}} ->
+          {:valid, [inner_changes | acc]}
+
+        {:valid, %Changeset{valid?: false, errors: errors}} ->
+          # The implementation is "all or nothing", so even if there where successful
+          # changes before, we reset them now since the whole changeset isn't valid.
+          {:invalid, errors}
+
+        {:invalid, %Changeset{valid?: false, errors: errors}} ->
+          {:invalid, errors ++ acc}
```

Comment:
I chose the syntax for simplicity. The size of the error list will be at most `lenght(:rules) * length(array)`, so even for a complicated form with 10 item rules and 50 list entries, this will be at most 500 elements. 

According to these benchmarks https://github.com/devonestes/fast-elixir#combining-lists-with--vs--code this will not be significant.